037-prototypes/proposal.md

```diff
@@ -0,0 +1,473 @@
+# Concourse Prototype Protocol
+
+This proposal outlines a small protocol for invoking conceptually related
+commands in a container image with a JSON-based request-response. This protocol
+is to be used as a foundation for both a new resource type interface and as a
+way to package and reuse arbitrary functionality (i.e. tasks).
+
+The protocol is versioned, starting at `1.0`.
+
+An implementation of this protocol is called a **prototype**. ([Why
+'prototype'?](#etymology)) Conceptually, a prototype handles **messages** sent
+to **objects** which are provided by the user as configuration or produced by
+the prototype itself in response to a prior message.
+
+For example, a `git` prototype may handle a `check` message against a
+repository object to produce commit objects, which then can then handle a `get`
+message to fetch the commit.
+
+An object's supported messages are discovered through an [Info
+request](#prototype-info). For example, the `git` prototype may support `get`
+against a commit object, but not against a repository object. The Info request
+is also a good place to perform validation.
```

Comment:
This section gives me another way to put my points in other comments, which is, I would rather "assume" things (perhaps with compiler time errors now) about my resources when I use a pipeline. And so you could say that all I want is to introduce a comparable generalization of pipelines, and say that pipelines still only work with the resource prototype-type.